Sources/SmithyXML/Writer/Writer.swift

```diff
@@ -0,0 +1,223 @@
+    }
+
+    public func writeTimestamp(_ value: Date?, format: TimestampFormat) throws {
+        guard let value else { detach(); return }
+        record(string: TimestampFormatter(format: format).string(from: value))
+    }
+
+    public func write<T: RawRepresentable>(_ value: T?) throws where T.RawValue == Int {
+        try write(value?.rawValue)
+    }
+
+    public func write<T: RawRepresentable>(_ value: T?) throws where T.RawValue == String {
+        try write(value?.rawValue)
+    }
+
+    public func writeMap<T>(
+        _ value: [String: T]?,
+        valueWritingClosure: WritingClosure<T, Writer>,
+        keyNodeInfo: NodeInfo,
+        valueNodeInfo: NodeInfo,
+        isFlattened: Bool
+    ) throws {
+        guard let value else { detach(); return }
+        if isFlattened {
+            guard let parent = self.parent else { return }
+            for (key, value) in value {
+                let entryWriter = parent[.init(element.name ?? "")]
+                try entryWriter[keyNodeInfo].write(key)
+                try valueWritingClosure(value, entryWriter[valueNodeInfo])
+            }
+            detach()
+        } else {
+            for (key, value) in value {
+                let entryWriter = self[.init("entry")]
+                try entryWriter[keyNodeInfo].write(key)
+                try valueWritingClosure(value, entryWriter[valueNodeInfo])
+            }
+        }
+    }
+
+    public func writeList<T>(
+        _ value: [T]?,
+        memberWritingClosure: WritingClosure<T, Writer>,
+        memberNodeInfo: NodeInfo,
+        isFlattened: Bool
+    ) throws {
+        guard let value else { detach(); return }
+        if isFlattened {
+            guard let parent = self.parent else { return }
+            let flattenedMemberNodeInfo = NodeInfo(
+                element.name ?? "",
+                location: memberNodeInfo.location,
+                namespace: memberNodeInfo.namespace
+            )
+            for member in value {
+                try memberWritingClosure(member, parent[flattenedMemberNodeInfo])
+            }
+            detach()
+        } else {
+            for member in value {
+                try memberWritingClosure(member, self[memberNodeInfo])
+            }
+        }
+    }
+
+    public func write(_ value: Data?) throws {
+        try write(value?.base64EncodedString())
+    }
+
+    public func detach() {
+        element.detach()
+    }
+
+    private func record(string: String?) {
+        guard let string, let key = nodeInfoPath.last else { detach(); return }
+        let xmlKey = nodeInfoPath.last
+        switch xmlKey?.kind ?? .element {
+        case .attribute:
+            guard let parent = element.parent as? XMLElement else { break }
+            let attribute = XMLNode(kind: .attribute)
+            attribute.name = key.name
+            attribute.stringValue = string
+            parent.addAttribute(attribute)
+            detach()
```

Comment:
could consider making this a helper function `private func addAttribute(name: String, value: String)`

Reply:
Done, though I actually made this into a custom initializer for `XMLElement`.